### PR TITLE
Fix : normalize collection

### PR DESCRIPTION
**Pull request by MLKiiwy**
In certain case collection are normalize into object (json_encode transform array( 1 => 'A' ) into { '1' : 'a' } instead of ['a'].

Here is the bug fix with the corresponding scenario failing on master.

### CODE REVIEW

**Comment by vincentchalamon** [comment]
Don't forget to create a PR on the 1.0 branch

**Comment by vincentchalamon** [comment]
:+1:

File: JsonLd/Serializer/ItemNormalizer.php [hunk 1]
```diff
@@ -152,7 +152,7 @@ public function normalize($object, $format = null, array $context = [])
                 $subResource = $this->resourceResolver->getResourceFromType($collectionType)
             ) {
                 $values = [];
-                foreach ($attributeValue as $index => $obj) {
+                foreach ($attributeValue->getValues() as $index => $obj) {
```
Comment (dunglas) [inline JsonLd/Serializer/ItemNormalizer.php]: I've maybe missed something but how can you be sure that the `getValues()` method exists? Isn't it something specific to Doctrine?

**Comment by dunglas** [comment]
Can't you fix your entity getter instead (to add the `getValues()` call)? If I understand well how it works it add an hard dependency to Doctrine Collections.

File: JsonLd/Serializer/ItemNormalizer.php (hunk 1 above)
Comment (MLKiiwy) [inline JsonLd/Serializer/ItemNormalizer.php]: The getValues method is from Collection interface of symfony. So every collection on symfony have it. It prevent the issue that array collection internal array can be not ordered properly.

If you want we can add : 
```php
$values = $attributeValue instanceof Collection ? $attributeValue->getValues() : $attributeValue;
```

Comment (dunglas) [inline JsonLd/Serializer/ItemNormalizer.php]: AFAIK there is no `Collection` interface in Symfony. Only in Doctrine collection. And this serializer can be used with other kind of `\Traversable` such as plain old arrays.

Comment (theofidry) [inline JsonLd/Serializer/ItemNormalizer.php]: same concern here, we're not always using Doctrine Collections ;)

**Comment by theofidry** [comment]
How do we do when we actually want to use hashmaps? In my case I don't really care as I've never had to use it with DunglasApiBundle, but it would be an issue to completely prevent this use case IMO.

**Comment by MLKiiwy** [comment]
@theofidry you're right ! I have an issue now with a relation that is indexed with a string so a hashmap ....

So this simple fix can't works ... 
@dunglas I suggest to modifiy the vendor php-property-info because the issue is to identify hash and arrays. In php hash == array but in json hash == object and array == array.

In https://github.com/dunglas/php-property-info/blob/master/src/PropertyInfo/Type.php you can identify a collection, but no difference between hash // array.
If we add a difference : isHash() for exemple => we can modify DoctrineExtractor to identify relation indexed by string.

What do you think ?

**Comment by vincentchalamon** [comment]
@dunglas @theofidry What do you think about @MLKiiwy question ?

**Comment by dunglas** [comment]
PHP Property Info is deprecated and has been merged into Symfony as the [Symfony PropertyInfo Component](http://symfony.com/blog/new-in-symfony-2-8-propertyinfo-component).

I don't think that this is the responsibility of this component to handle such differences. What about just having smart getter and setters updating array correctly?

**Comment by theofidry** [comment]
Well, I would say he's suggestion makes sense but would still be tricky because you can still use an hashmap with int values :/ The best thing IMO is to either reset the keys in the array setter (or remover if you have one) or return all the values in the getter (but likely more costly performance wise):

```php
public function setMyArray(array $arr)
{
    $this->myArray = array_values($arr);
}

public function addToMyArray($val)
{
    $this->myArray[] = $val;
}

public function removeFromMyArray($val)
{
    $key = array_search($val, $this->myArray, true);

    if (false === $key) {
        return;
    }

    if (0 === $key) {
        array_shift($this->myArray);

        return;
    }

    unset($this->myArray[$key]);
    $this->myArray = array_values($this->myArray);
}
```

File: JsonLd/Serializer/ItemNormalizer.php (hunk 1 above)
Comment (sroze) [inline JsonLd/Serializer/ItemNormalizer.php]: We could use Doctrine's `Collection` interface as it's in `doctrine/common` and it not related/dependent at all on any database-related stuff.

**Comment by sroze** [comment]
@MLKiiwy any news about this PR?

File: JsonLd/Serializer/ItemNormalizer.php (hunk 1 above)
Comment (teohhanhui) [inline JsonLd/Serializer/ItemNormalizer.php]:
```php
$values = $attributeValue instanceof \Traversable ? iterator_to_array($attributeValue, false) : array_values($attributeValue);
```

**Comment by teohhanhui** [comment]
@dunglas
> Can't you fix your entity getter instead (to add the getValues() call)? If I understand well how it works it add a hard dependency to Doctrine Collections to (IMO) fix a Doctrine bug/edge case.

As I've mentioned in https://github.com/dunglas/DunglasApiBundle/issues/285#issuecomment-152983128 it's not a "Doctrine edge case". If we know it's a to-many relation we should normalize it to numerically-indexed non-sparse array.

@theofidry
> How do we do when we actually want to use hashmaps?

For a relation? You mean like with Doctrine's "indexBy" option? I'd think that's a responsibility of the mapping metadata layer to provide such information...

**Comment by dunglas** [comment]
Fixed in 2.0.

**Comment by teohhanhui** [comment]
How was this fixed?

**Comment by dunglas** [comment]
Maybe I closed to quickly.

**Comment by MLKiiwy** [comment]
I'm closing this one, need to test in 2.0 if the issue is still here

**Pull request closed without merging**